src/System.Management.Automation/engine/CommandCompletion/CompletionHelpers.cs

```diff
@@ -21,36 +21,65 @@ internal static class CompletionHelpers
         internal static IEnumerable<CompletionResult> GetMatchingResults(
             string wordToComplete,
             IEnumerable<string> possibleCompletionValues,
             Func<string, string> toolTipMapping = null,
+            Func<string, string> listItemTextMapping = null,
             CompletionResultType resultType = CompletionResultType.Text)
         {
             string quote = HandleDoubleAndSingleQuote(ref wordToComplete);
-            var pattern = WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase);
 
             foreach (string value in possibleCompletionValues)
             {
-                if (pattern.IsMatch(value))
+                if (IsMatch(value, wordToComplete))
                 {
                     string completionText = QuoteCompletionText(value, quote);
+                    string toolTip = toolTipMapping?.Invoke(value) ?? value;
+                    string listItemText = listItemTextMapping?.Invoke(value) ?? value;
 
-                    string listItemText = value;
-
-                    yield return new CompletionResult(
-                        completionText,
-                        listItemText,
-                        resultType,
-                        toolTip: toolTipMapping is null
-                            ? listItemText
-                            : toolTipMapping(value));
+                    yield return new CompletionResult(completionText, listItemText, resultType, toolTip);
                 }
             }
         }
 
+        /// <summary>
+        /// Checks if the given value matches the specified word or pattern.
+        /// </summary>
+        /// <param name="value">The input string to evaluate.</param>
+        /// <param name="wordToComplete">The word or pattern to compare against.</param>
+        /// <returns>
+        /// <c>true</c> if the value matches the normalized word (case-insensitively) or the wildcard pattern; otherwise, <c>false</c>.
+        /// </returns>
+        /// <remarks>
+        /// This method normalizes word to complete by replacing its line endings with backticks 
+        /// and unescaping special characters. 
+        /// It first does a a case-insensitive literal prefix match.
+        /// If the literal match fails, the input value is then evaluated against this case-insensitive wildcard pattern.
+        /// To prevent any exceptions when constructing wildcard pattern with existing wildcard characters,
+        /// the word to complete is normalized by escaping wildcard characters first.
+        /// </remarks>
+        internal static bool IsMatch(string value, string wordToComplete)
+        {
+            string normalizedWord = WildcardPattern.Unescape(wordToComplete.ReplaceLineEndings("`"));
+
+            bool match = value.StartsWith(normalizedWord, StringComparison.OrdinalIgnoreCase);
+
+            if (!match)
+            {
+                normalizedWord = WildcardPattern.Escape(wordToComplete);
```

Comment:
Oh, this can already be considered a breaking change - the original code handled wildcard characters, now we are blocking this for all completers.
It might be worth making IsMatch a delegate already - default one without escaping.

Reply:
Yep sounds good. I have made an `escapeStrategy` delegate which by default does not escape: 33ca0a73cecbb05762a94f6bdf361da17f0fb02b.